src/interpreter/mod.rs

```diff
@@ -734,16 +734,12 @@ impl<'a, 'tcx> EvalContext<'a, 'tcx> {
             ty::TyRef(_, ty::TypeAndMut { ty, .. }) |
             ty::TyRawPtr(ty::TypeAndMut { ty, .. }) |
             ty::TyBox(ty) => {
-                if self.type_is_sized(ty) {
-                    assert_eq!(field_index, 0);
-                    Ok(self.tcx.mk_imm_ptr(self.tcx.types.never))
-                } else {
-                    match (field_index, &ty.sty) {
-                        (1, &ty::TySlice(_)) => Ok(self.tcx.types.usize),
-                        (1, &ty::TyTrait(_)) |
-                        (0, _) => Ok(self.tcx.mk_imm_ptr(self.tcx.types.never)),
-                        _ => bug!("invalid fat pointee type"),
-                    }
+                assert!(!self.type_is_sized(ty));
+                match (field_index, &self.tcx.struct_tail(ty).sty) {
+                    (1, &ty::TySlice(_)) => Ok(self.tcx.types.usize),
```

Comment:
This should also mention `TyStr`... *However*, neither `[T]` nor `str`'s length fields can be used for the optimization.
It's probably okay to keep it? I'm not sure. Also, the `assert!(!self.type_is_sized(ty));` can be removed since you use `struct_tail`, all you have to do is match on `TyTrait` / `TySlice` / `TyStr` specifically for the pointer field.